sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala

```diff
@@ -158,18 +180,28 @@ object SchemaPruning extends Rule[LogicalPlan] {
 
   /**
    * Builds a pruned logical relation from the output of the output relation and the schema of the
-   * pruned base relation.
+   * pruned base relation
    */
   private def buildPrunedRelation(
       outputRelation: LogicalRelation,
-      prunedBaseRelation: HadoopFsRelation) = {
+      prunedBaseRelation: HadoopFsRelation,
+      prunedMetadataSchema: StructType) = {
     val prunedOutput = getPrunedOutput(outputRelation.output, prunedBaseRelation.schema)
```

Comment:
how about
```
val prunedOutput = getPrunedOutput(outputRelation.output, prunedBaseRelation.schema)
val finalOutput = prunedOutput.map {
  case MetadataAttribute(attr) => attr.withDataType(prunedMetadataSchema)
  case other => other
}
outputRelation.copy(relation = prunedBaseRelation, output = finalOutput)
```

Reply:
actually, this can be even simpler. We only care about the final schema here, as `getPrunedOutput` is basically generating new attributes from the final schema and storing expr IDs from `relation.output`
```
val prunedOutput = getPrunedOutput(outputRelation.output, finalSchema)
outputRelation.copy(relation = prunedBaseRelation, output = prunedOutput)
```

Reply:
thanks! great idea, also need to slightly change `getPrunedOutput` to keep `metadata` of the attribute.